exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/DrillFunctionRegistry.java

```diff
@@ -218,4 +302,141 @@ private void registerOperatorsWithoutInference(DrillOperatorTable operatorTable)
       }
     }
   }
+
+  /**
+   * Function registry holder. Stores function implementations by jar name, function name.
+   */
+  private class GenericRegistryHolder<T, U> {
+    private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+    private final AutoCloseableLock readLock = new AutoCloseableLock(readWriteLock.readLock());
+    private final AutoCloseableLock writeLock = new AutoCloseableLock(readWriteLock.writeLock());
+
+    // jar name, Map<function name, List<signature>
+    private final Map<T, Map<T, List<T>>> jars;
+
+    // function name, Map<signature, function holder>
```

Comment:
T appears to be a String. functions is a map of signature to map of Strings to FuncHolders. This is crying out for an explanation. How is the signature represented? Classic Java encoding? Something else? What does the Map hold? What is the key in the inner map? Function name? Jar name? Meaning will become clear by studying the code, but much time could be saved by explaining the structure in a comment.